From 8d4fc742177471def4b7f42422077b8962b8a2a3 Mon Sep 17 00:00:00 2001
From: John Lindgren <john@jlindgren.net>
Date: Thu, 27 Jun 2024 19:17:38 -0400
Subject: [PATCH] xdg: support xdg-shell v3 with popup repositioning

See https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3514
which added support on the wlroots side.

We now re-run popup positioning (for both xdg-shell and layer-shell
popups) when the "reposition" event is received. This allows popups that
change size (such as qmpanel's applications menu) to be positioned
correctly.

xdg-shell v3 also gives the compositor some additional "hints" for popup
positioning (reactive, parent_size, and parent_configure_serial) which
are available but we don't make use of currently.

Upstream-Status: Backport [https://github.com/labwc/labwc/commit/a98f2635ea60e416891cbc381a1a3c0d6a3478de]
---
 include/layers.h |  1 +
 src/layers.c     | 13 +++++++++++++
 src/xdg-popup.c  | 12 ++++++++++++
 src/xdg.c        |  2 +-
 4 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/include/layers.h b/include/layers.h
index 52583035..e4d671f3 100644
--- a/include/layers.h
+++ b/include/layers.h
@@ -31,6 +31,7 @@ struct lab_layer_popup {
 	struct wl_listener commit;
 	struct wl_listener destroy;
 	struct wl_listener new_popup;
+	struct wl_listener reposition;
 };
 
 void layers_init(struct server *server);
diff --git a/src/layers.c b/src/layers.c
index 20e9e683..cfaa7274 100644
--- a/src/layers.c
+++ b/src/layers.c
@@ -260,6 +260,7 @@ popup_handle_destroy(struct wl_listener *listener, void *data)
 		wl_container_of(listener, popup, destroy);
 	wl_list_remove(&popup->destroy.link);
 	wl_list_remove(&popup->new_popup.link);
+	wl_list_remove(&popup->reposition.link);
 
 	/* Usually already removed unless there was no commit at all */
 	if (popup->commit.notify) {
@@ -285,6 +286,15 @@ popup_handle_commit(struct wl_listener *listener, void *data)
 	}
 }
 
+static void
+popup_handle_reposition(struct wl_listener *listener, void *data)
+{
+	struct lab_layer_popup *popup =
+		wl_container_of(listener, popup, reposition);
+	wlr_xdg_popup_unconstrain_from_box(popup->wlr_popup,
+		&popup->output_toplevel_sx_box);
+}
+
 static void popup_handle_new_popup(struct wl_listener *listener, void *data);
 
 static struct lab_layer_popup *
@@ -310,6 +320,9 @@ create_popup(struct wlr_xdg_popup *wlr_popup, struct wlr_scene_tree *parent)
 	popup->commit.notify = popup_handle_commit;
 	wl_signal_add(&wlr_popup->base->surface->events.commit, &popup->commit);
 
+	popup->reposition.notify = popup_handle_reposition;
+	wl_signal_add(&wlr_popup->events.reposition, &popup->reposition);
+
 	return popup;
 }
 
diff --git a/src/xdg-popup.c b/src/xdg-popup.c
index 65c13e8d..50288d07 100644
--- a/src/xdg-popup.c
+++ b/src/xdg-popup.c
@@ -19,6 +19,7 @@ struct xdg_popup {
 	struct wl_listener commit;
 	struct wl_listener destroy;
 	struct wl_listener new_popup;
+	struct wl_listener reposition;
 };
 
 static void
@@ -50,6 +51,7 @@ handle_xdg_popup_destroy(struct wl_listener *listener, void *data)
 	struct xdg_popup *popup = wl_container_of(listener, popup, destroy);
 	wl_list_remove(&popup->destroy.link);
 	wl_list_remove(&popup->new_popup.link);
+	wl_list_remove(&popup->reposition.link);
 
 	/* Usually already removed unless there was no commit at all */
 	if (popup->commit.notify) {
@@ -72,6 +74,13 @@ handle_xdg_popup_commit(struct wl_listener *listener, void *data)
 	}
 }
 
+static void
+handle_xdg_popup_reposition(struct wl_listener *listener, void *data)
+{
+	struct xdg_popup *popup = wl_container_of(listener, popup, reposition);
+	popup_unconstrain(popup);
+}
+
 static void
 popup_handle_new_xdg_popup(struct wl_listener *listener, void *data)
 {
@@ -103,6 +112,9 @@ xdg_popup_create(struct view *view, struct wlr_xdg_popup *wlr_popup)
 	popup->commit.notify = handle_xdg_popup_commit;
 	wl_signal_add(&wlr_popup->base->surface->events.commit, &popup->commit);
 
+	popup->reposition.notify = handle_xdg_popup_reposition;
+	wl_signal_add(&wlr_popup->events.reposition, &popup->reposition);
+
 	/*
 	 * We must add xdg popups to the scene graph so they get rendered. The
 	 * wlroots scene graph provides a helper for this, but to use it we must
diff --git a/src/xdg.c b/src/xdg.c
index dc9acc91..f14660df 100644
--- a/src/xdg.c
+++ b/src/xdg.c
@@ -13,7 +13,7 @@
 #include "window-rules.h"
 #include "workspaces.h"
 
-#define LAB_XDG_SHELL_VERSION (2)
+#define LAB_XDG_SHELL_VERSION (3)
 #define CONFIGURE_TIMEOUT_MS 100
 
 static struct xdg_toplevel_view *
-- 
2.34.1

